searcher: Multiline search

keegancsmith opened the pull request.

keegancsmith commented on cmd/searcher/search/matcher.go:
```diff
@@ -216,65 +215,102 @@ func (rg *readerGrep) Find(zf *store.ZipFile, f *store.SrcFile) (matches []proto
+		lineLimitHit := len(locs) == maxOffsets
+		for _, match := range locs {
+			start, end := match[0], match[1]
+			lineStart := bytes.LastIndex(fileMatchBuf[:start], []byte{'\n'})
```
you should be able to bound this somehow? ie we would have an older lineStart or lineEnd from an earlier match. That should avoid the need to look at the whole of fileMatchBuf upto start.

attfarhan commented:
@keegancsmith here is the first few levels of the profile on the benchmark with the biggest delta (searches for `.*`):
![image](https://user-images.githubusercontent.com/16265452/61149362-cee97700-a495-11e9-879d-60977cde8ebe.png)
RuneCount takes a long time, which I already reduced by following your suggestion in the comment above (using the previous line's match as the first index so we don't count from the beginning of the file). The new things that look problematic seem to be `bytes.SplitN`, which we do when we split results with new lines, and then `regexp.FindAllIndex`, which is probably worse since we match on the entire file.

attfarhan commented:
on master, the profile for that same benchmark:

![image](https://user-images.githubusercontent.com/16265452/61150992-a499b880-a499-11e9-82a0-85246ee04cfb.png)

attfarhan commented:
So it looks like `RuneCount` is really the big problem here

attfarhan commented:
Ah ok, I got rid of that huge RuneCount bloat by also making RuneCount start at the starting byte of the line for the first line of a match also.

attfarhan commented:
New benchstat output: 

```
BytesToLowerASCII/short-8                            10.8ns ± 3%    12.2ns ± 5%    +13.06%  (p=0.000 n=10+10)
BytesToLowerASCII/pangram-8                          5.85ns ± 3%    7.06ns ± 2%    +20.57%  (p=0.000 n=10+9)
BytesToLowerASCII/8k-8                                538ns ± 1%     789ns ± 3%    +46.72%  (p=0.000 n=10+10)
BytesToLowerASCII/8k-misaligned-8                     537ns ± 0%     784ns ± 6%    +46.01%  (p=0.000 n=8+10)
ConcurrentFind_large_fixed-8                         9.52ms ± 3%   10.05ms ± 4%     +5.59%  (p=0.000 n=9+10)
ConcurrentFind_large_fixed_casesensitive-8           7.37ms ± 4%    7.83ms ± 4%     +6.25%  (p=0.000 n=9+10)
ConcurrentFind_large_re_dotstar-8                    18.3ms ± 2%   206.5ms ± 6%  +1025.54%  (p=0.000 n=10+10)
ConcurrentFind_large_re_common-8                     11.4ms ± 5%    10.4ms ± 5%     -9.50%  (p=0.000 n=9+10)
ConcurrentFind_large_re_anchor-8                      134ms ± 6%     153ms ± 2%    +14.74%  (p=0.000 n=10+10)
ConcurrentFind_large_path/path_only-8                 522µs ± 7%     502µs ± 3%       ~     (p=0.079 n=10+9)
ConcurrentFind_large_path/content_only-8             6.59ms ± 1%    9.10ms ± 6%    +38.08%  (p=0.000 n=10+9)
ConcurrentFind_large_path/both_path_and_content-8    6.76ms ± 3%    9.45ms ± 4%    +39.67%  (p=0.000 n=10+9)
ConcurrentFind_small_fixed-8                          219µs ± 4%     175µs ± 3%    -20.30%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed_casesensitive-8            114µs ± 2%      98µs ±33%    -13.95%  (p=0.004 n=9+9)
ConcurrentFind_small_re_dotstar-8                    2.02ms ± 1%   13.70ms ± 7%   +578.57%  (p=0.000 n=9+9)
ConcurrentFind_small_re_common-8                      328µs ± 3%     255µs ± 3%    -22.17%  (p=0.000 n=9+10)
ConcurrentFind_small_re_anchor-8                     2.15ms ± 5%    2.48ms ± 7%    +15.39%  (p=0.000 n=10+9)

name                                               old alloc/op   new alloc/op   delta
ConcurrentFind_large_fixed-8                         4.95MB ± 1%    5.03MB ± 1%     +1.74%  (p=0.000 n=9+10)
ConcurrentFind_large_fixed_casesensitive-8           11.2kB ± 0%    51.8kB ± 2%   +360.69%  (p=0.000 n=8+10)
ConcurrentFind_large_re_dotstar-8                    33.2MB ± 0%   613.0MB ± 0%  +1746.84%  (p=0.000 n=10+10)
ConcurrentFind_large_re_common-8                     2.73MB ± 1%   16.99MB ± 0%   +522.86%  (p=0.000 n=10+10)
ConcurrentFind_large_re_anchor-8                     2.51MB ± 2%   15.60MB ± 0%   +520.21%  (p=0.000 n=10+10)
ConcurrentFind_large_path/path_only-8                  630B ± 1%      632B ± 1%       ~     (p=0.285 n=10+10)
ConcurrentFind_large_path/content_only-8              111kB ± 8%     484kB ± 1%   +335.26%  (p=0.000 n=10+10)
ConcurrentFind_large_path/both_path_and_content-8     109kB ± 7%     483kB ± 1%   +341.36%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed-8                          474kB ± 5%     428kB ±12%     -9.79%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed_casesensitive-8           5.10kB ± 1%   16.80kB ± 0%   +229.17%  (p=0.000 n=8+10)
ConcurrentFind_small_re_dotstar-8                    2.90MB ± 0%   31.54MB ± 0%   +987.76%  (p=0.000 n=10+10)
ConcurrentFind_small_re_common-8                     60.6kB ± 1%   324.9kB ± 0%   +436.18%  (p=0.000 n=10+10)
ConcurrentFind_small_re_anchor-8                     19.9kB ± 3%   101.8kB ± 1%   +411.86%  (p=0.000 n=10+10)

name                                               old allocs/op  new allocs/op  delta
ConcurrentFind_large_fixed-8                            210 ± 2%       220 ± 2%     +4.69%  (p=0.000 n=9+10)
ConcurrentFind_large_fixed_casesensitive-8              115 ± 1%       127 ± 0%    +10.15%  (p=0.000 n=10+8)
ConcurrentFind_large_re_dotstar-8                      278k ± 0%      970k ± 0%   +249.05%  (p=0.000 n=10+9)
ConcurrentFind_large_re_common-8                      24.0k ± 0%     26.2k ± 0%     +9.23%  (p=0.000 n=10+10)
ConcurrentFind_large_re_anchor-8                      22.7k ± 0%     24.6k ± 0%     +8.40%  (p=0.000 n=10+10)
ConcurrentFind_large_path/path_only-8                  16.0 ± 0%      16.0 ± 0%       ~     (all equal)
ConcurrentFind_large_path/content_only-8                648 ± 0%       693 ± 0%     +6.93%  (p=0.000 n=10+10)
ConcurrentFind_large_path/both_path_and_content-8       648 ± 1%       692 ± 1%     +6.89%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed-8                           91.2 ± 2%      83.0 ± 2%     -8.99%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed_casesensitive-8             60.0 ± 0%      63.0 ± 0%     +5.00%  (p=0.000 n=10+10)
ConcurrentFind_small_re_dotstar-8                     21.7k ± 0%     48.5k ± 0%   +123.28%  (p=0.000 n=10+10)
ConcurrentFind_small_re_common-8                        486 ± 0%       523 ± 0%     +7.55%  (p=0.000 n=10+10)
ConcurrentFind_small_re_anchor-8                        208 ± 0%       232 ± 0%    +11.54%  (p=0.000 n=10+8)
```

attfarhan commented:
New profile:
[prof.svg.zip](https://github.com/sourcegraph/sourcegraph/files/3387736/prof.svg.zip)

attfarhan commented:
New profile for benchmark `large_re_dotstar`:
[pprof002.svg.zip](https://github.com/sourcegraph/sourcegraph/files/3388295/pprof002.svg.zip)

keegancsmith commented:
Ok we can work on the bytes.SplitN.

BTW I rebased and squashed the changes. You accidentally included some binary outputs and changes to go.mod/go.sum.

keegancsmith commented:
Sorry I didn’t have a lot of time to look at this today / my brain is a bit fried when I did. I am dedicating tomorrow morning to this PR and the symbol zoekt work. Expect detailed feedback / some code. My approach will to be to play with the code to try understand what is going on.

My initial instinct is two fold for what to investigate:
- Correctness - maybe this implementation is incorrect -> it is doing way more work. That could explain the regression.
- bytes.SplitN - We don’t have to do bytes.SplitN, we could be more incremental in our approach. (ie scan for `\n` until we don’t find it).

You may want to explore that today. However, from our earlier benchmarks I am 75% confident this approach will work. Tomorrow you should have a much higher confidence answer :)

keegancsmith commented:
I just pushed a wip of something which is performant. When looking at the profiles above its clear that `bytes.SplitN` introduced too much slowness. Instead I introduced `appendMatches` which is like bytes.SplitN, except tailored to our usecase of appending matches. I have no idea about correctness / etc, but this approach seems very fruitful.

Additionally I believe this is faster than our old searcher implementation because we avoid line parsing, and then rely on optimized `bytes.Count` to quickly count what line we on. Before we did a lot more work per line, even if it had no matches. Updated benchmarks:

```
name                                               old time/op    new time/op    delta
BytesToLowerASCII/short-8                            7.29ns ± 1%   10.30ns ± 1%   +41.35%  (p=0.002 n=10+4)
BytesToLowerASCII/pangram-8                          4.68ns ± 1%    5.37ns ± 5%   +14.80%  (p=0.004 n=8+4)
BytesToLowerASCII/8k-8                                420ns ± 4%     474ns ± 6%   +12.90%  (p=0.002 n=10+4)
BytesToLowerASCII/8k-misaligned-8                     429ns ± 4%     488ns ± 3%   +13.53%  (p=0.002 n=10+4)
ConcurrentFind_large_fixed-8                         8.28ms ± 9%    8.15ms ±13%      ~     (p=1.000 n=10+4)
ConcurrentFind_large_fixed_casesensitive-8           6.60ms ± 2%    6.32ms ± 2%    -4.18%  (p=0.004 n=10+4)
ConcurrentFind_large_re_dotstar-8                    17.0ms ± 2%    71.6ms ± 2%  +321.00%  (p=0.002 n=10+4)
ConcurrentFind_large_re_common-8                     10.3ms ± 4%     5.0ms ± 5%   -51.57%  (p=0.003 n=9+4)
ConcurrentFind_large_re_anchor-8                      118ms ± 4%     122ms ± 5%      ~     (p=0.054 n=10+4)
ConcurrentFind_large_path/path_only-8                 414µs ± 2%     415µs ± 4%      ~     (p=0.604 n=9+4)
ConcurrentFind_large_path/content_only-8             5.99ms ± 6%    7.09ms ± 6%   +18.37%  (p=0.002 n=10+4)
ConcurrentFind_large_path/both_path_and_content-8    6.26ms ± 3%    7.62ms ± 6%   +21.66%  (p=0.002 n=10+4)
ConcurrentFind_small_fixed-8                          183µs ± 8%     129µs ± 7%   -29.69%  (p=0.002 n=10+4)
ConcurrentFind_small_fixed_casesensitive-8           96.6µs ± 1%    65.7µs ± 1%   -31.97%  (p=0.003 n=9+4)
ConcurrentFind_small_re_dotstar-8                    1.81ms ± 3%    3.47ms ± 2%   +91.06%  (p=0.002 n=10+4)
ConcurrentFind_small_re_common-8                      248µs ± 3%     112µs ± 2%   -54.68%  (p=0.003 n=9+4)
ConcurrentFind_small_re_anchor-8                     1.81ms ± 4%    2.01ms ± 2%   +11.21%  (p=0.002 n=10+4)

name                                               old alloc/op   new alloc/op   delta
ConcurrentFind_large_fixed-8                         5.02MB ± 1%    5.08MB ± 1%    +1.27%  (p=0.004 n=8+4)
ConcurrentFind_large_fixed_casesensitive-8           12.4kB ± 9%    11.3kB ± 3%    -8.65%  (p=0.020 n=10+4)
ConcurrentFind_large_re_dotstar-8                    33.1MB ± 0%    63.3MB ± 0%   +91.20%  (p=0.002 n=10+4)
ConcurrentFind_large_re_common-8                     2.73MB ± 1%    1.90MB ± 0%   -30.15%  (p=0.002 n=10+4)
ConcurrentFind_large_re_anchor-8                     2.51MB ± 1%    1.81MB ± 2%   -27.89%  (p=0.003 n=9+4)
ConcurrentFind_large_path/path_only-8                  628B ± 0%      644B ± 1%    +2.55%  (p=0.002 n=10+4)
ConcurrentFind_large_path/content_only-8              107kB ± 5%      78kB ± 1%   -27.16%  (p=0.002 n=10+4)
ConcurrentFind_large_path/both_path_and_content-8     108kB ± 3%      78kB ± 2%   -27.20%  (p=0.002 n=10+4)
ConcurrentFind_small_fixed-8                          458kB ± 8%     430kB ± 9%      ~     (p=0.240 n=10+4)
ConcurrentFind_small_fixed_casesensitive-8           5.10kB ± 1%    4.29kB ± 1%   -15.88%  (p=0.002 n=10+4)
ConcurrentFind_small_re_dotstar-8                    2.89MB ± 1%    3.71MB ± 0%   +28.27%  (p=0.002 n=10+4)
ConcurrentFind_small_re_common-8                     60.6kB ± 0%    40.4kB ± 0%   -33.41%  (p=0.002 n=10+4)
ConcurrentFind_small_re_anchor-8                     20.0kB ± 3%    16.9kB ± 2%   -15.39%  (p=0.002 n=10+4)

name                                               old allocs/op  new allocs/op  delta
ConcurrentFind_large_fixed-8                            215 ± 4%       228 ± 3%    +6.16%  (p=0.004 n=10+4)
ConcurrentFind_large_fixed_casesensitive-8              116 ± 0%       112 ± 0%    -3.45%  (p=0.000 n=8+4)
ConcurrentFind_large_re_dotstar-8                      278k ± 0%      746k ± 0%  +168.63%  (p=0.003 n=9+4)
ConcurrentFind_large_re_common-8                      24.0k ± 0%     20.5k ± 0%   -14.62%  (p=0.002 n=10+4)
ConcurrentFind_large_re_anchor-8                      22.7k ± 0%     19.4k ± 0%   -14.46%  (p=0.002 n=10+4)
ConcurrentFind_large_path/path_only-8                  16.0 ± 0%      16.0 ± 0%      ~     (all equal)
ConcurrentFind_large_path/content_only-8                647 ± 0%       538 ± 0%   -16.81%  (p=0.002 n=10+4)
ConcurrentFind_large_path/both_path_and_content-8       647 ± 0%       538 ± 0%   -16.89%  (p=0.002 n=10+4)
ConcurrentFind_small_fixed-8                           90.6 ± 3%      82.2 ± 3%    -9.22%  (p=0.002 n=10+4)
ConcurrentFind_small_fixed_casesensitive-8             60.0 ± 0%      58.0 ± 0%    -3.33%  (p=0.000 n=10+4)
ConcurrentFind_small_re_dotstar-8                     21.7k ± 0%     37.3k ± 0%   +71.62%  (p=0.003 n=9+4)
ConcurrentFind_small_re_common-8                        486 ± 0%       417 ± 0%   -14.20%  (p=0.000 n=10+4)
ConcurrentFind_small_re_anchor-8                        208 ± 0%       199 ± 0%    -4.33%  (p=0.000 n=10+4)
```

ijt commented:
@keegancsmith it looks like a many of the benchmarks are saying the new implementation is slower, especially `ConcurrentFind_large_re_dotstar-8`. What gives you the impression that it's faster?

keegancsmith commented:
> @keegancsmith it looks like a many of the benchmarks are saying the new implementation is slower, especially `ConcurrentFind_large_re_dotstar-8`. What gives you the impression that it's faster?

@ijt the only cases it is slower on is `.*` as a pattern. This is a degenerative case due to how much it matches. So I believe is acceptable to be slower on. For the other more realistic cases we have improved performance. I need to investigate it further, but we can probably improve .* as well.

attfarhan commented:
@keegancsmith wow, this new implementation is a huge improvement 🎉 . 

But, like @ijt, I'm curious about the benchmark results too. It looks like there are a few other cases besides the `.*` case that have worse perf. These seem like they would be realistic cases that we should care about as well, or are these also special cases that are OK to regress on? 

```
ConcurrentFind_large_path/content_only-8             5.99ms ± 6%    7.09ms ± 6%   +18.37%  (p=0.002 n=10+4)
ConcurrentFind_large_path/both_path_and_content-8    6.26ms ± 3%    7.62ms ± 6%   +21.66%  (p=0.002 n=10+4)
ConcurrentFind_small_re_anchor-8                     1.81ms ± 4%    2.01ms ± 2%   +11.21%  (p=0.002 n=10+4)
```

attfarhan commented:
cc @keegancsmith ran the benchmarks again without doing stuff on my computer as you suggested. Did it a number of times just in case, and saw the results are mostly the same. The first one showed less degradation but the others still showed `large_path/content_only` and `large_path/both_path_and_content` were worse in terms of perf
```
name                                               old time/op    new time/op    delta
BytesToLowerASCII/short-8                            11.1ns ± 3%    12.0ns ± 7%    +8.42%  (p=0.000 n=9+10)
BytesToLowerASCII/pangram-8                          6.74ns ± 6%    6.59ns ± 6%      ~     (p=0.160 n=10+10)
BytesToLowerASCII/8k-8                                682ns ± 4%     745ns ± 7%    +9.16%  (p=0.000 n=10+10)
BytesToLowerASCII/8k-misaligned-8                     720ns ± 8%     720ns ± 2%      ~     (p=0.562 n=10+9)
ConcurrentFind_large_fixed-8                         10.3ms ± 2%    10.1ms ± 1%      ~     (p=0.093 n=9+8)
ConcurrentFind_large_fixed_casesensitive-8           8.79ms ±15%    8.10ms ± 6%      ~     (p=0.063 n=10+10)
ConcurrentFind_large_re_dotstar-8                    22.3ms ±16%    84.3ms ± 3%  +277.93%  (p=0.000 n=10+10)
ConcurrentFind_large_re_common-8                     13.5ms ± 8%     5.6ms ± 2%   -58.55%  (p=0.000 n=10+8)
ConcurrentFind_large_re_anchor-8                      171ms ±18%     133ms ± 3%   -22.16%  (p=0.000 n=10+9)
ConcurrentFind_large_path/path_only-8                 626µs ±15%     522µs ± 4%   -16.55%  (p=0.000 n=10+9)
ConcurrentFind_large_path/content_only-8             8.24ms ± 6%    8.58ms ± 2%    +4.13%  (p=0.001 n=10+8)
ConcurrentFind_large_path/both_path_and_content-8    8.61ms ± 7%    8.74ms ± 2%      ~     (p=0.218 n=10+10)
ConcurrentFind_small_fixed-8                          247µs ± 8%     164µs ± 6%   -33.64%  (p=0.000 n=9+10)
ConcurrentFind_small_fixed_casesensitive-8            152µs ± 7%      83µs ± 4%   -45.37%  (p=0.000 n=10+10)
ConcurrentFind_small_re_dotstar-8                    2.45ms ±18%    3.92ms ± 3%   +59.86%  (p=0.000 n=10+9)
ConcurrentFind_small_re_common-8                      407µs ±14%     138µs ± 2%   -66.10%  (p=0.000 n=10+9)
ConcurrentFind_small_re_anchor-8                     2.46ms ±10%    2.03ms ± 3%   -17.15%  (p=0.000 n=10+10)

name                                               old alloc/op   new alloc/op   delta
BytesToLowerASCII/short-8                             0.00B          0.00B           ~     (all equal)
BytesToLowerASCII/pangram-8                           0.00B          0.00B           ~     (all equal)
BytesToLowerASCII/8k-8                                0.00B          0.00B           ~     (all equal)
BytesToLowerASCII/8k-misaligned-8                     0.00B          0.00B           ~     (all equal)
ConcurrentFind_large_fixed-8                         4.98MB ± 1%    4.86MB ± 1%    -2.52%  (p=0.000 n=10+9)
ConcurrentFind_large_fixed_casesensitive-8           12.8kB ± 3%    10.2kB ± 0%   -20.37%  (p=0.000 n=10+10)
ConcurrentFind_large_re_dotstar-8                    33.2MB ± 0%    63.2MB ± 0%   +90.53%  (p=0.000 n=10+10)
ConcurrentFind_large_re_common-8                     2.69MB ± 0%    1.76MB ± 0%   -34.81%  (p=0.000 n=10+10)
ConcurrentFind_large_re_anchor-8                     2.49MB ± 2%    1.66MB ± 1%   -33.14%  (p=0.000 n=10+9)
ConcurrentFind_large_path/path_only-8                  634B ± 1%      640B ± 4%      ~     (p=0.337 n=10+10)
ConcurrentFind_large_path/content_only-8             92.8kB ± 6%    62.7kB ± 3%   -32.39%  (p=0.000 n=10+10)
ConcurrentFind_large_path/both_path_and_content-8    93.6kB ± 5%    62.8kB ± 4%   -32.97%  (p=0.000 n=10+9)
ConcurrentFind_small_fixed-8                          365kB ±30%     396kB ±13%      ~     (p=0.481 n=10+10)
ConcurrentFind_small_fixed_casesensitive-8           5.04kB ± 2%    4.25kB ± 0%   -15.66%  (p=0.000 n=10+8)
ConcurrentFind_small_re_dotstar-8                    2.80MB ± 1%    3.57MB ± 0%   +27.27%  (p=0.000 n=10+10)
ConcurrentFind_small_re_common-8                     58.2kB ± 1%    39.4kB ± 0%   -32.30%  (p=0.000 n=10+10)
ConcurrentFind_small_re_anchor-8                     20.2kB ± 4%    16.9kB ± 3%   -16.62%  (p=0.000 n=10+10)

name                                               old allocs/op  new allocs/op  delta
BytesToLowerASCII/short-8                              0.00           0.00           ~     (all equal)
BytesToLowerASCII/pangram-8                            0.00           0.00           ~     (all equal)
BytesToLowerASCII/8k-8                                 0.00           0.00           ~     (all equal)
BytesToLowerASCII/8k-misaligned-8                      0.00           0.00           ~     (all equal)
ConcurrentFind_large_fixed-8                            210 ± 2%       187 ± 2%   -11.00%  (p=0.000 n=9+9)
ConcurrentFind_large_fixed_casesensitive-8              116 ± 0%       102 ± 1%   -11.72%  (p=0.000 n=10+10)
ConcurrentFind_large_re_dotstar-8                      278k ± 0%      745k ± 0%  +168.22%  (p=0.000 n=10+10)
ConcurrentFind_large_re_common-8                      23.9k ± 0%     19.4k ± 0%   -18.95%  (p=0.000 n=10+9)
ConcurrentFind_large_re_anchor-8                      22.7k ± 0%     18.4k ± 0%   -18.98%  (p=0.000 n=10+10)
ConcurrentFind_large_path/path_only-8                  16.0 ± 0%      16.0 ± 0%      ~     (all equal)
ConcurrentFind_large_path/content_only-8                640 ± 0%       510 ± 0%   -20.30%  (p=0.000 n=10+10)
ConcurrentFind_large_path/both_path_and_content-8       640 ± 0%       510 ± 0%   -20.27%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed-8                           80.1 ± 8%      74.5 ± 2%    -6.99%  (p=0.003 n=10+10)
ConcurrentFind_small_fixed_casesensitive-8             60.0 ± 0%      56.0 ± 0%    -6.67%  (p=0.000 n=10+10)
ConcurrentFind_small_re_dotstar-8                     21.7k ± 0%     37.2k ± 0%   +71.29%  (p=0.000 n=10+6)
ConcurrentFind_small_re_common-8                        484 ± 0%       394 ± 0%   -18.66%  (p=0.000 n=10+10)
ConcurrentFind_small_re_anchor-8                        208 ± 0%       178 ± 0%   -14.42%  (p=0.000 n=10+10)
```

```
name                                               old time/op    new time/op    delta
BytesToLowerASCII/short-8                            11.2ns ± 4%    12.0ns ± 7%    +7.52%  (p=0.000 n=10+10)
BytesToLowerASCII/pangram-8                          6.64ns ± 6%    6.59ns ± 6%      ~     (p=0.424 n=10+10)
BytesToLowerASCII/8k-8                                652ns ± 1%     745ns ± 7%   +14.24%  (p=0.000 n=8+10)
BytesToLowerASCII/8k-misaligned-8                     677ns ± 5%     720ns ± 2%    +6.37%  (p=0.000 n=10+9)
ConcurrentFind_large_fixed-8                         10.4ms ± 3%    10.1ms ± 1%    -2.33%  (p=0.006 n=10+8)
ConcurrentFind_large_fixed_casesensitive-8           8.18ms ± 2%    8.10ms ± 6%      ~     (p=0.436 n=10+10)
ConcurrentFind_large_re_dotstar-8                    20.7ms ± 2%    84.3ms ± 3%  +307.74%  (p=0.000 n=10+10)
ConcurrentFind_large_re_common-8                     12.7ms ± 3%     5.6ms ± 2%   -55.98%  (p=0.000 n=9+8)
ConcurrentFind_large_re_anchor-8                      141ms ± 8%     133ms ± 3%    -5.30%  (p=0.000 n=10+9)
ConcurrentFind_large_path/path_only-8                 504µs ± 5%     522µs ± 4%    +3.60%  (p=0.015 n=8+9)
ConcurrentFind_large_path/content_only-8             7.21ms ± 9%    8.58ms ± 2%   +19.04%  (p=0.000 n=10+8)
ConcurrentFind_large_path/both_path_and_content-8    7.60ms ±11%    8.74ms ± 2%   +15.13%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed-8                          206µs ± 3%     164µs ± 6%   -20.25%  (p=0.000 n=9+10)
ConcurrentFind_small_fixed_casesensitive-8            122µs ± 2%      83µs ± 4%   -32.06%  (p=0.000 n=9+10)
ConcurrentFind_small_re_dotstar-8                    1.94ms ± 4%    3.92ms ± 3%  +102.59%  (p=0.000 n=10+9)
ConcurrentFind_small_re_common-8                      351µs ± 4%     138µs ± 2%   -60.62%  (p=0.000 n=10+9)
ConcurrentFind_small_re_anchor-8                     2.16ms ± 1%    2.03ms ± 3%    -5.76%  (p=0.000 n=9+10)

name                                               old alloc/op   new alloc/op   delta
BytesToLowerASCII/short-8                             0.00B          0.00B           ~     (all equal)
BytesToLowerASCII/pangram-8                           0.00B          0.00B           ~     (all equal)
BytesToLowerASCII/8k-8                                0.00B          0.00B           ~     (all equal)
BytesToLowerASCII/8k-misaligned-8                     0.00B          0.00B           ~     (all equal)
ConcurrentFind_large_fixed-8                         4.92MB ± 1%    4.86MB ± 1%    -1.32%  (p=0.001 n=10+9)
ConcurrentFind_large_fixed_casesensitive-8           12.7kB ± 4%    10.2kB ± 0%   -20.04%  (p=0.000 n=10+10)
ConcurrentFind_large_re_dotstar-8                    33.2MB ± 0%    63.2MB ± 0%   +90.60%  (p=0.000 n=10+10)
ConcurrentFind_large_re_common-8                     2.69MB ± 0%    1.76MB ± 0%   -34.78%  (p=0.000 n=10+10)
ConcurrentFind_large_re_anchor-8                     2.51MB ± 1%    1.66MB ± 1%   -33.87%  (p=0.000 n=10+9)
ConcurrentFind_large_path/path_only-8                  632B ± 1%      640B ± 4%      ~     (p=0.210 n=9+10)
ConcurrentFind_large_path/content_only-8             93.1kB ± 5%    62.7kB ± 3%   -32.65%  (p=0.000 n=10+10)
ConcurrentFind_large_path/both_path_and_content-8    95.4kB ± 3%    62.8kB ± 4%   -34.21%  (p=0.000 n=10+9)
ConcurrentFind_small_fixed-8                          481kB ± 2%     396kB ±13%   -17.76%  (p=0.000 n=8+10)
ConcurrentFind_small_fixed_casesensitive-8           5.04kB ± 1%    4.25kB ± 0%   -15.63%  (p=0.000 n=10+8)
ConcurrentFind_small_re_dotstar-8                    2.82MB ± 1%    3.57MB ± 0%   +26.61%  (p=0.000 n=10+10)
ConcurrentFind_small_re_common-8                     58.6kB ± 2%    39.4kB ± 0%   -32.68%  (p=0.000 n=10+10)
ConcurrentFind_small_re_anchor-8                     19.6kB ± 4%    16.9kB ± 3%   -14.09%  (p=0.000 n=10+10)

name                                               old allocs/op  new allocs/op  delta
BytesToLowerASCII/short-8                              0.00           0.00           ~     (all equal)
BytesToLowerASCII/pangram-8                            0.00           0.00           ~     (all equal)
BytesToLowerASCII/8k-8                                 0.00           0.00           ~     (all equal)
BytesToLowerASCII/8k-misaligned-8                      0.00           0.00           ~     (all equal)
ConcurrentFind_large_fixed-8                            207 ± 2%       187 ± 2%    -9.61%  (p=0.000 n=9+9)
ConcurrentFind_large_fixed_casesensitive-8              116 ± 0%       102 ± 1%   -11.72%  (p=0.000 n=10+10)
ConcurrentFind_large_re_dotstar-8                      278k ± 0%      745k ± 0%  +168.23%  (p=0.000 n=9+10)
ConcurrentFind_large_re_common-8                      23.9k ± 0%     19.4k ± 0%   -18.95%  (p=0.000 n=10+9)
ConcurrentFind_large_re_anchor-8                      22.7k ± 0%     18.4k ± 0%   -19.00%  (p=0.000 n=10+10)
ConcurrentFind_large_path/path_only-8                  16.0 ± 0%      16.0 ± 0%      ~     (all equal)
ConcurrentFind_large_path/content_only-8                640 ± 0%       510 ± 0%   -20.33%  (p=0.000 n=10+10)
ConcurrentFind_large_path/both_path_and_content-8       640 ± 0%       510 ± 0%   -20.33%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed-8                           86.6 ± 1%      74.5 ± 2%   -14.00%  (p=0.000 n=8+10)
ConcurrentFind_small_fixed_casesensitive-8             60.0 ± 0%      56.0 ± 0%    -6.67%  (p=0.000 n=10+10)
ConcurrentFind_small_re_dotstar-8                     21.7k ± 0%     37.2k ± 0%   +71.28%  (p=0.000 n=9+6)
ConcurrentFind_small_re_common-8                        485 ± 0%       394 ± 0%   -18.70%  (p=0.000 n=10+10)
ConcurrentFind_small_re_anchor-8                        208 ± 0%       178 ± 0%   -14.42%  (p=0.000 n=10+10)
```

```
BytesToLowerASCII/short-8                            11.0ns ± 5%    12.0ns ± 7%    +9.48%  (p=0.000 n=10+10)
BytesToLowerASCII/pangram-8                          6.49ns ± 6%    6.59ns ± 6%      ~     (p=0.118 n=10+10)
BytesToLowerASCII/8k-8                                670ns ± 5%     745ns ± 7%   +11.18%  (p=0.000 n=10+10)
BytesToLowerASCII/8k-misaligned-8                     673ns ± 4%     720ns ± 2%    +7.03%  (p=0.000 n=10+9)
ConcurrentFind_large_fixed-8                         10.8ms ±10%    10.1ms ± 1%    -6.12%  (p=0.034 n=10+8)
ConcurrentFind_large_fixed_casesensitive-8           8.05ms ± 4%    8.10ms ± 6%      ~     (p=0.739 n=10+10)
ConcurrentFind_large_re_dotstar-8                    20.7ms ± 3%    84.3ms ± 3%  +306.27%  (p=0.000 n=10+10)
ConcurrentFind_large_re_common-8                     12.9ms ± 6%     5.6ms ± 2%   -56.61%  (p=0.000 n=10+8)
ConcurrentFind_large_re_anchor-8                      138ms ± 3%     133ms ± 3%    -3.79%  (p=0.000 n=10+9)
ConcurrentFind_large_path/path_only-8                 522µs ± 8%     522µs ± 4%      ~     (p=0.905 n=10+9)
ConcurrentFind_large_path/content_only-8             7.34ms ± 5%    8.58ms ± 2%   +16.89%  (p=0.000 n=9+8)
ConcurrentFind_large_path/both_path_and_content-8    7.53ms ± 3%    8.74ms ± 2%   +16.17%  (p=0.000 n=9+10)
ConcurrentFind_small_fixed-8                          196µs ± 5%     164µs ± 6%   -16.19%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed_casesensitive-8            124µs ± 5%      83µs ± 4%   -32.81%  (p=0.000 n=9+10)
ConcurrentFind_small_re_dotstar-8                   22.0ms ±141%     3.9ms ± 3%      ~     (p=0.156 n=10+9)
ConcurrentFind_small_re_common-8                      396µs ±30%     138µs ± 2%   -65.17%  (p=0.000 n=10+9)
ConcurrentFind_small_re_anchor-8                     2.54ms ± 6%    2.03ms ± 3%   -20.04%  (p=0.000 n=10+10)

name                                               old alloc/op   new alloc/op   delta
BytesToLowerASCII/short-8                             0.00B          0.00B           ~     (all equal)
BytesToLowerASCII/pangram-8                           0.00B          0.00B           ~     (all equal)
BytesToLowerASCII/8k-8                                0.00B          0.00B           ~     (all equal)
BytesToLowerASCII/8k-misaligned-8                     0.00B          0.00B           ~     (all equal)
ConcurrentFind_large_fixed-8                         4.98MB ± 1%    4.86MB ± 1%    -2.42%  (p=0.000 n=10+9)
ConcurrentFind_large_fixed_casesensitive-8           12.7kB ± 5%    10.2kB ± 0%   -20.34%  (p=0.000 n=10+10)
ConcurrentFind_large_re_dotstar-8                    33.1MB ± 0%    63.2MB ± 0%   +90.69%  (p=0.000 n=8+10)
ConcurrentFind_large_re_common-8                     2.69MB ± 0%    1.76MB ± 0%   -34.78%  (p=0.000 n=10+10)
ConcurrentFind_large_re_anchor-8                     2.49MB ± 1%    1.66MB ± 1%   -33.27%  (p=0.000 n=9+9)
ConcurrentFind_large_path/path_only-8                  630B ± 1%      640B ± 4%      ~     (p=0.066 n=9+10)
ConcurrentFind_large_path/content_only-8             94.1kB ± 5%    62.7kB ± 3%   -33.39%  (p=0.000 n=10+10)
ConcurrentFind_large_path/both_path_and_content-8    93.2kB ± 3%    62.8kB ± 4%   -32.66%  (p=0.000 n=10+9)
ConcurrentFind_small_fixed-8                          391kB ±18%     396kB ±13%      ~     (p=0.796 n=10+10)
ConcurrentFind_small_fixed_casesensitive-8           5.03kB ± 2%    4.25kB ± 0%   -15.53%  (p=0.000 n=10+8)
ConcurrentFind_small_re_dotstar-8                    2.80MB ± 1%    3.57MB ± 0%   +27.44%  (p=0.000 n=10+10)
ConcurrentFind_small_re_common-8                     58.6kB ± 1%    39.4kB ± 0%   -32.76%  (p=0.000 n=10+10)
ConcurrentFind_small_re_anchor-8                     19.9kB ± 4%    16.9kB ± 3%   -15.19%  (p=0.000 n=10+10)

name                                               old allocs/op  new allocs/op  delta
BytesToLowerASCII/short-8                              0.00           0.00           ~     (all equal)
BytesToLowerASCII/pangram-8                            0.00           0.00           ~     (all equal)
BytesToLowerASCII/8k-8                                 0.00           0.00           ~     (all equal)
BytesToLowerASCII/8k-misaligned-8                      0.00           0.00           ~     (all equal)
ConcurrentFind_large_fixed-8                            209 ± 2%       187 ± 2%   -10.44%  (p=0.000 n=10+9)
ConcurrentFind_large_fixed_casesensitive-8              116 ± 0%       102 ± 1%   -11.72%  (p=0.000 n=10+10)
ConcurrentFind_large_re_dotstar-8                      278k ± 0%      745k ± 0%  +168.24%  (p=0.000 n=10+10)
ConcurrentFind_large_re_common-8                      23.9k ± 0%     19.4k ± 0%   -18.95%  (p=0.000 n=10+9)
ConcurrentFind_large_re_anchor-8                      22.7k ± 0%     18.4k ± 0%   -19.00%  (p=0.000 n=9+10)
ConcurrentFind_large_path/path_only-8                  16.0 ± 0%      16.0 ± 0%      ~     (all equal)
ConcurrentFind_large_path/content_only-8                640 ± 0%       510 ± 0%   -20.37%  (p=0.000 n=8+10)
ConcurrentFind_large_path/both_path_and_content-8       640 ± 0%       510 ± 0%   -20.24%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed-8                           82.0 ± 4%      74.5 ± 2%    -9.15%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed_casesensitive-8             60.0 ± 0%      56.0 ± 0%    -6.67%  (p=0.000 n=10+10)
ConcurrentFind_small_re_dotstar-8                     21.7k ± 0%     37.2k ± 0%   +71.30%  (p=0.000 n=10+6)
ConcurrentFind_small_re_common-8                        485 ± 0%       394 ± 0%   -18.76%  (p=0.000 n=10+10)
ConcurrentFind_small_re_anchor-8                        208 ± 0%       178 ± 0%   -14.42%  (p=0.000 n=10+10)
```

```
name                                               old time/op    new time/op    delta
BytesToLowerASCII/short-8                            11.0ns ± 5%    11.7ns ± 3%    +6.35%  (p=0.000 n=10+9)
BytesToLowerASCII/pangram-8                          6.49ns ± 6%    6.85ns ± 6%    +5.63%  (p=0.002 n=10+10)
BytesToLowerASCII/8k-8                                670ns ± 5%     810ns ± 2%   +20.93%  (p=0.000 n=10+9)
BytesToLowerASCII/8k-misaligned-8                     673ns ± 4%     772ns ± 7%   +14.79%  (p=0.000 n=10+10)
ConcurrentFind_large_fixed-8                         10.8ms ±10%    10.4ms ± 5%      ~     (p=0.123 n=10+10)
ConcurrentFind_large_fixed_casesensitive-8           8.05ms ± 4%    8.15ms ± 3%      ~     (p=0.393 n=10+10)
ConcurrentFind_large_re_dotstar-8                    20.7ms ± 3%    85.6ms ± 1%  +312.49%  (p=0.000 n=10+10)
ConcurrentFind_large_re_common-8                     12.9ms ± 6%     5.4ms ± 6%   -58.35%  (p=0.000 n=10+10)
ConcurrentFind_large_re_anchor-8                      138ms ± 3%     137ms ± 4%      ~     (p=0.481 n=10+10)
ConcurrentFind_large_path/path_only-8                 522µs ± 8%     529µs ± 8%      ~     (p=0.529 n=10+10)
ConcurrentFind_large_path/content_only-8             7.34ms ± 5%    8.64ms ± 4%   +17.72%  (p=0.000 n=9+10)
ConcurrentFind_large_path/both_path_and_content-8    7.53ms ± 3%    8.74ms ± 2%   +16.10%  (p=0.000 n=9+10)
ConcurrentFind_small_fixed-8                          196µs ± 5%     190µs ±46%      ~     (p=0.083 n=10+8)
ConcurrentFind_small_fixed_casesensitive-8            124µs ± 5%     654µs ±16%  +428.18%  (p=0.000 n=9+8)
ConcurrentFind_small_re_dotstar-8                   22.0ms ±141%     4.3ms ± 8%      ~     (p=0.156 n=10+9)
ConcurrentFind_small_re_common-8                      396µs ±30%     152µs ± 5%   -61.71%  (p=0.000 n=10+10)
ConcurrentFind_small_re_anchor-8                     2.54ms ± 6%    2.26ms ± 4%   -11.38%  (p=0.000 n=10+10)

name                                               old alloc/op   new alloc/op   delta
BytesToLowerASCII/short-8                             0.00B          0.00B           ~     (all equal)
BytesToLowerASCII/pangram-8                           0.00B          0.00B           ~     (all equal)
BytesToLowerASCII/8k-8                                0.00B          0.00B           ~     (all equal)
BytesToLowerASCII/8k-misaligned-8                     0.00B          0.00B           ~     (all equal)
ConcurrentFind_large_fixed-8                         4.98MB ± 1%    5.00MB ± 1%      ~     (p=0.190 n=10+10)
ConcurrentFind_large_fixed_casesensitive-8           12.7kB ± 5%    11.4kB ± 5%   -10.91%  (p=0.000 n=10+10)
ConcurrentFind_large_re_dotstar-8                    33.1MB ± 0%    63.3MB ± 0%   +90.87%  (p=0.000 n=8+10)
ConcurrentFind_large_re_common-8                     2.69MB ± 0%    1.80MB ± 1%   -33.25%  (p=0.000 n=10+10)
ConcurrentFind_large_re_anchor-8                     2.49MB ± 1%    1.68MB ± 3%   -32.49%  (p=0.000 n=9+10)
ConcurrentFind_large_path/path_only-8                  630B ± 1%      632B ± 1%      ~     (p=0.321 n=9+10)
ConcurrentFind_large_path/content_only-8             94.1kB ± 5%    69.4kB ± 4%   -26.26%  (p=0.000 n=10+10)
ConcurrentFind_large_path/both_path_and_content-8    93.2kB ± 3%    67.9kB ± 3%   -27.17%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed-8                          391kB ±18%     417kB ±24%      ~     (p=0.165 n=10+10)
ConcurrentFind_small_fixed_casesensitive-8           5.03kB ± 2%    4.54kB ± 3%    -9.75%  (p=0.000 n=10+9)
ConcurrentFind_small_re_dotstar-8                    2.80MB ± 1%    3.52MB ± 0%   +25.71%  (p=0.000 n=10+9)
ConcurrentFind_small_re_common-8                     58.6kB ± 1%    42.0kB ± 0%   -28.46%  (p=0.000 n=10+10)
ConcurrentFind_small_re_anchor-8                     19.9kB ± 4%    17.1kB ± 2%   -14.06%  (p=0.000 n=10+9)

name                                               old allocs/op  new allocs/op  delta
BytesToLowerASCII/short-8                              0.00           0.00           ~     (all equal)
BytesToLowerASCII/pangram-8                            0.00           0.00           ~     (all equal)
BytesToLowerASCII/8k-8                                 0.00           0.00           ~     (all equal)
BytesToLowerASCII/8k-misaligned-8                      0.00           0.00           ~     (all equal)
ConcurrentFind_large_fixed-8                            209 ± 2%       198 ± 2%    -5.17%  (p=0.000 n=10+10)
ConcurrentFind_large_fixed_casesensitive-8              116 ± 0%       103 ± 0%   -11.21%  (p=0.000 n=10+10)
ConcurrentFind_large_re_dotstar-8                      278k ± 0%      745k ± 0%  +168.24%  (p=0.000 n=10+9)
ConcurrentFind_large_re_common-8                      23.9k ± 0%     19.4k ± 0%   -18.87%  (p=0.000 n=10+10)
ConcurrentFind_large_re_anchor-8                      22.7k ± 0%     18.4k ± 0%   -18.95%  (p=0.000 n=9+10)
ConcurrentFind_large_path/path_only-8                  16.0 ± 0%      16.0 ± 0%      ~     (all equal)
ConcurrentFind_large_path/content_only-8                640 ± 0%       513 ± 0%   -19.92%  (p=0.000 n=8+10)
ConcurrentFind_large_path/both_path_and_content-8       640 ± 0%       512 ± 0%   -19.94%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed-8                           82.0 ± 4%      77.8 ± 6%    -5.12%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed_casesensitive-8             60.0 ± 0%      56.0 ± 0%    -6.67%  (p=0.000 n=10+9)
ConcurrentFind_small_re_dotstar-8                     21.7k ± 0%     37.2k ± 0%   +71.28%  (p=0.000 n=10+10)
ConcurrentFind_small_re_common-8                        485 ± 0%       395 ± 0%   -18.56%  (p=0.000 n=10+10)
ConcurrentFind_small_re_anchor-8                        208 ± 0%       179 ± 0%   -13.94%  (p=0.000 n=10+9)
```

attfarhan commented:
@keegancsmith latest benchmarks run (with latest commit on this PR), results are quite nice:
```
BytesToLowerASCII/short-8                            10.8ns ± 5%    12.1ns ± 4%  +11.56%  (p=0.000 n=8+10)
BytesToLowerASCII/pangram-8                          6.95ns ± 5%    6.84ns ± 4%     ~     (p=0.196 n=10+10)
BytesToLowerASCII/8k-8                                691ns ± 5%     690ns ± 6%     ~     (p=0.928 n=10+10)
BytesToLowerASCII/8k-misaligned-8                     707ns ± 6%     679ns ± 5%   -4.00%  (p=0.019 n=10+10)
ConcurrentFind_large_fixed-8                         10.7ms ± 5%    10.5ms ± 6%     ~     (p=0.353 n=10+10)
ConcurrentFind_large_fixed_casesensitive-8           8.38ms ± 2%    8.27ms ± 3%     ~     (p=0.278 n=9+10)
ConcurrentFind_large_re_dotstar-8                    20.9ms ± 3%    21.2ms ± 3%     ~     (p=0.105 n=10+10)
ConcurrentFind_large_re_common-8                     13.0ms ± 3%     6.2ms ±46%  -52.13%  (p=0.000 n=10+10)
ConcurrentFind_large_re_anchor-8                      141ms ± 2%     171ms ±43%     ~     (p=0.481 n=10+10)
ConcurrentFind_large_path/path_only-8                 527µs ± 5%     550µs ± 7%   +4.33%  (p=0.034 n=10+8)
ConcurrentFind_large_path/content_only-8             7.35ms ± 6%   11.97ms ±42%  +62.81%  (p=0.000 n=10+10)
ConcurrentFind_large_path/both_path_and_content-8    7.81ms ± 3%    8.71ms ± 6%  +11.59%  (p=0.000 n=9+10)
ConcurrentFind_small_fixed-8                          213µs ± 2%     164µs ± 4%  -22.98%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed_casesensitive-8            127µs ± 5%      84µs ± 2%  -33.73%  (p=0.000 n=10+9)
ConcurrentFind_small_re_dotstar-8                    1.98ms ± 5%    1.67ms ± 2%  -15.54%  (p=0.000 n=10+9)
ConcurrentFind_small_re_common-8                      351µs ± 5%     140µs ± 2%  -60.03%  (p=0.000 n=10+10)
ConcurrentFind_small_re_anchor-8                     2.32ms ± 6%    2.11ms ± 3%   -8.65%  (p=0.000 n=9+9)

name                                               old alloc/op   new alloc/op   delta
BytesToLowerASCII/short-8                             0.00B          0.00B          ~     (all equal)
BytesToLowerASCII/pangram-8                           0.00B          0.00B          ~     (all equal)
BytesToLowerASCII/8k-8                                0.00B          0.00B          ~     (all equal)
BytesToLowerASCII/8k-misaligned-8                     0.00B          0.00B          ~     (all equal)
ConcurrentFind_large_fixed-8                         5.00MB ± 0%    4.97MB ± 1%   -0.54%  (p=0.000 n=10+10)
ConcurrentFind_large_fixed_casesensitive-8           12.7kB ± 5%    10.1kB ± 0%  -20.13%  (p=0.000 n=10+8)
ConcurrentFind_large_re_dotstar-8                    33.2MB ± 0%    22.7MB ± 0%  -31.48%  (p=0.000 n=10+10)
ConcurrentFind_large_re_common-8                     2.69MB ± 0%    1.72MB ± 0%  -35.85%  (p=0.000 n=10+10)
ConcurrentFind_large_re_anchor-8                     2.49MB ± 3%    1.65MB ± 3%  -33.90%  (p=0.000 n=10+10)
ConcurrentFind_large_path/path_only-8                  630B ± 1%      640B ± 2%   +1.50%  (p=0.002 n=10+9)
ConcurrentFind_large_path/content_only-8             93.1kB ± 5%    62.6kB ± 2%  -32.74%  (p=0.000 n=10+9)
ConcurrentFind_large_path/both_path_and_content-8    94.3kB ± 3%    62.5kB ± 3%  -33.68%  (p=0.000 n=9+10)
ConcurrentFind_small_fixed-8                          474kB ± 6%     435kB ± 6%   -8.27%  (p=0.000 n=9+10)
ConcurrentFind_small_fixed_casesensitive-8           5.03kB ± 1%    4.27kB ± 1%  -15.06%  (p=0.000 n=10+10)
ConcurrentFind_small_re_dotstar-8                    2.81MB ± 0%    1.91MB ± 2%  -32.22%  (p=0.000 n=10+10)
ConcurrentFind_small_re_common-8                     58.3kB ± 1%    39.5kB ± 1%  -32.31%  (p=0.000 n=10+10)
ConcurrentFind_small_re_anchor-8                     19.7kB ± 2%    16.5kB ± 2%  -16.20%  (p=0.000 n=10+9)

name                                               old allocs/op  new allocs/op  delta
BytesToLowerASCII/short-8                              0.00           0.00          ~     (all equal)
BytesToLowerASCII/pangram-8                            0.00           0.00          ~     (all equal)
BytesToLowerASCII/8k-8                                 0.00           0.00          ~     (all equal)
BytesToLowerASCII/8k-misaligned-8                      0.00           0.00          ~     (all equal)
ConcurrentFind_large_fixed-8                            209 ± 2%       190 ± 1%   -9.44%  (p=0.000 n=10+8)
ConcurrentFind_large_fixed_casesensitive-8              116 ± 0%       102 ± 0%  -12.07%  (p=0.000 n=10+8)
ConcurrentFind_large_re_dotstar-8                      278k ± 0%      215k ± 0%  -22.65%  (p=0.000 n=10+10)
ConcurrentFind_large_re_common-8                      23.9k ± 0%     19.0k ± 0%  -20.77%  (p=0.000 n=10+10)
ConcurrentFind_large_re_anchor-8                      22.7k ± 0%     18.0k ± 0%  -20.52%  (p=0.000 n=9+10)
ConcurrentFind_large_path/path_only-8                  16.0 ± 0%      16.0 ± 0%     ~     (all equal)
ConcurrentFind_large_path/content_only-8                640 ± 0%       510 ± 0%  -20.34%  (p=0.000 n=10+10)
ConcurrentFind_large_path/both_path_and_content-8       640 ± 0%       510 ± 0%  -20.42%  (p=0.000 n=10+10)
ConcurrentFind_small_fixed-8                           85.8 ± 1%      76.3 ± 2%  -11.05%  (p=0.000 n=9+10)
ConcurrentFind_small_fixed_casesensitive-8             60.0 ± 0%      56.0 ± 0%   -6.67%  (p=0.000 n=10+10)
ConcurrentFind_small_re_dotstar-8                     21.7k ± 0%     16.9k ± 0%  -22.23%  (p=0.000 n=10+10)
ConcurrentFind_small_re_common-8                        485 ± 0%       394 ± 0%  -18.71%  (p=0.000 n=10+10)
ConcurrentFind_small_re_anchor-8                        208 ± 0%       178 ± 0%  -14.42%  (p=0.000 n=10+9)
```

keegancsmith commented on cmd/searcher/search/matcher.go:
```diff
@@ -289,7 +286,7 @@ func appendMatches(matches []protocol.LineMatch, fileBuf []byte, matchLineBuf []
 			LineNumber:       lineNumber,
 			OffsetAndLengths: [][2]int{{offset, length}},
-			LimitHit:         false, // TODO: should we remove/ignore this param now that we don't scan for matches by line?
+			LimitHit:         false, // We will always return false for this field since we no longer limit the number of offsets per line.
```
we need to add a limit in. Otherwise you could quite easily OOM us by searching something like `.`

attfarhan replied:
This is not a concern anymore as we discussed on Slack, right?

keegancsmith commented:
> @keegancsmith latest benchmarks run (with latest commit on this PR), results are quite nice:

Great! Will take a look at this PR.

keegancsmith commented on cmd/searcher/search/search_test.go:
```diff
@@ -36,7 +36,8 @@ func TestSearch(t *testing.T) {
 
-Hello world example in go`,
+Hello world example in go
+`,
```
this test before hand was actually exposing a bug in the implementation. Files don't always end with `\n`, so we need to fix this issue.

keegancsmith replied:
I pushed a fix for this.

attfarhan replied:
@keegancsmith I just saw this comment + commit, but I don't think this commit fixes it. 
In the snippet you added:
```
lineEnd := end
		if idx := bytes.Index(fileMatchBuf[end:], []byte{'\n'}); idx >= 0 {
			lineEnd = end + idx
		}			}
```
If we are in a file that does _not_ have a newline at the end, `lineEnd` stops at `end` which is incorrect.

keegancsmith commented:
We need to add test cases for searching multiline. Especially corner cases. But this is close to merging!

keegancsmith commented on cmd/searcher/search/matcher.go:
```diff
@@ -211,70 +203,99 @@ func (rg *readerGrep) Find(zf *store.ZipFile, f *store.SrcFile) (matches []proto
-		if idx < first[0] {
-			continue
+		// TODO double check this makes sense with the conditionals below
+		lineStart := lastLineStartIndex + bytes.LastIndex(fileMatchBuf[lastLineStartIndex:start], []byte{'\n'})
```
I think this is a source of slowdown in the current implementation. Think about a file which is just one line `aaaaaa...`, `a` 10,000 times. Then if you search for `a`, for each match you will have

```
start = 0, 1, 2, 3 ... 10,000
lastLineStartIndex = 0
```

So it degrades into `O(n)` per lookup, which overall is `O(n^2)` algorithm. I think line should actually be something which looks for a newline in `fileMatchBuf[lastStart:start]`. Then if it fails to find a newline, it can use `lastLineStartIndex`. Then I believe this will be `O(n)`.

I'll see if I can fix this and keep the tests passing. But we need to add a bunch of test cases for corner cases on multiline searching.

keegancsmith replied:
ok I think I just pushed a commit which solves this. Tests still pass :)

keegancsmith commented:
I pushed a change which may improve the benchmarks for anything which searches over a long lines.

keegancsmith commented:
PR is at https://github.com/sourcegraph/sourcegraph/pull/5165
